src/pending_connection/connect.rs

```diff
@@ -1,208 +1,261 @@
-                    flags: SrtShakeFlags::TSBPDSND | SrtShakeFlags::TSBPDRCV, // TODO: the reference implementation sets a lot more of these, research
-                    latency: tsbpd_latency,
-                })),
-                ext_km: None,
-                // ext_km: self.crypto.as_mut().map(|manager| {
-                //     SrtControlPacket::KeyManagerRequest(SrtKeyMessage {
-                //         pt: 2,       // TODO: what is this
-                //         sign: 8_233, // TODO: again
-                //         keki: 0,
-                //         cipher: CipherType::CTR,
-                //         auth: 0,
-                //         se: 2,
-                //         salt: Vec::from(manager.salt()),
-                //         even_key: Some(manager.wrap_key().unwrap()),
-                //         odd_key: None,
-                //         wrap_data: [0; 8],
-                //     })
-                // }),
-                ext_config: None,
-            },
-            ..hs_info
-        }),
-    });
+}
 
-    sock.send((pack.clone(), remote)).await?;
+#[derive(Clone)]
+pub enum ConnectState {
+    Configured(SeqNumber),
+    InductionResponseWait(Packet),
+    ConclusionResponseWait(Packet),
+    Connected(ConnectionSettings),
+}
 
-    loop {
-        let (packet, from) = select! {
-            _ = send_interval.tick().fuse() => {sock.send((pack.clone(), remote)).await?; continue},
-            res = get_packet(sock).fuse() => res?
-        };
-        if let Packet::Control(ControlPacket {
-            dest_sockid,
-            control_type:
-                ControlTypes::Handshake(
-                    info @ HandshakeControlInfo {
+impl ConnectState {
+    pub fn new() -> ConnectState {
+        Configured(rand::random())
+    }
+}
+
+#[derive(Debug)]
+pub enum ConnectError {
+    //#[error("Expected Control packet, expected: {0} found: {1}")]
+    ControlExpected(ShakeType, DataPacket),
+    //#[error("Expected Handshake packet, expected: {0} found: {1}")]
+    HandshakeExpected(ShakeType, ControlTypes),
+    //#[error("Expected Induction (1) packet, found: {0}")]
+    InductionExpected(HandshakeControlInfo),
+    //#[error("Expected packets from different host, expected: {0} found: {1}")]
+    UnexpectedHost(SocketAddr, SocketAddr),
+    //#[error("Expected Conclusion (-1) packet, found: {0}")]
+    ConclusionExpected(HandshakeControlInfo),
+    //#[error("Unsupported protocol version, expected: v5 found v{0}")]
+    UnsupportedProtocolVersion(u32),
+    // TODO: why don't we validate the cookie on responses
+    //#[error("Received invalid cookie handshake from [address], expected: {0} found {1}")]
+    //InvalidHandshakeCookie(i32, i32),
+    // TODO: why don't we validate we have an SRT response
+    //#[error("Expected SRT handshake request in conclusion handshake, found {0}")]
+    //SrtHandshakeExpected(HandshakeControlInfo),
+}
+
+pub struct Connect(ConnectConfiguration, ConnectState);
```

Comment:
Same for here as above with with respect to publicity. 

Also, I have a slight preference for using structs here instead of a tuple, i think `.field` is a bit easier than `.0`. But I'm willing to be convinced either way on this.

Reply:
I actually agree, when I have less than three values in a struct I'll prefer a tuple, but then, I'll switch to struct when I see patterns of direct field access instead of restructuring.

Reply:
I actually agree, when I have less than three values in a struct I'll prefer a tuple, but then, I'll switch to struct when I see patterns of direct field access instead of de-structuring.